server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java

```diff
@@ -76,54 +76,46 @@ private static class Visitor extends SymbolVisitor<Context, Void> {
         public Void visitFunction(Function function, Context context) {
             String functionName = function.name();
             switch (functionName) {
-                case AndOperator.NAME:
-                    for (Symbol arg : function.arguments()) {
-                        arg.accept(this, context);
+                case NotPredicate.NAME -> {
+                }
+                case OrOperator.NAME -> {
+                    context.isHashJoinPossible = false;
+                }
+                case EqOperator.NAME -> {
+                    List<Symbol> arguments = function.arguments();
+
+                    var left = arguments.get(0);
+                    var leftContext = new Context();
+                    left.accept(this, leftContext);
+
+                    var right = arguments.get(1);
+                    var rightContext = new Context();
+                    right.accept(this, rightContext);
+
+                    if (leftContext.relations.size() == 1 &&
+                        rightContext.relations.size() == 1 &&
+                        !leftContext.relations.equals(rightContext.relations)) {
+                        context.isHashJoinPossible = true;
```

Comment:
```suggestion
                    context.isHashJoinPossible = true;
                  } else {
                      return null;
                  }
```

Reply:
We can exit early here as well.

Reply:
See comment here: https://github.com/crate/crate/pull/15687#discussion_r1530254255 about setting the flag before exiting.